rust/lance/src/dataset/take.rs

```diff
@@ -553,6 +558,49 @@ impl TakeBuilder {
     }
 }
 
+fn take_struct_array(array: &StructArray, indices: &UInt64Array) -> Result<Arc<dyn Array>> {
+    let nulls = array.nulls().map(|nulls| {
+        let is_valid = indices.iter().map(|index| {
+            if let Some(index) = index {
+                nulls.is_valid(index.to_usize().unwrap())
+            } else {
+                false
+            }
+        });
+        NullBuffer::new(BooleanBuffer::new(
+            Buffer::from_iter(is_valid),
+            0,
+            indices.len(),
+        ))
+    });
+
+    if array.fields().is_empty() {
+        return Ok(Arc::new(StructArray::new_empty_fields(
+            indices.len(),
+            nulls,
+        )));
+    }
+
+    let arrays = array
+        .columns()
+        .iter()
+        .map(|array| {
+            let array = match array.data_type() {
+                arrow::datatypes::DataType::Struct(_) => {
+                    take_struct_array(array.as_struct(), indices)?
+                }
+                _ => arrow_select::take::take(array, indices, None)?,
+            };
+            Ok(array)
+        })
+        .collect::<Result<Vec<_>>>()?;
+    Ok(Arc::new(StructArray::new(
+        array.fields().clone(),
+        arrays,
+        nulls,
+    )))
```

Comment:
Minor nit: we could probably just return `StructArray` here instead of `Arc<dyn Array>` since the next thing you do is convert it to a `RecordBatch` anyways.